File: internal/findings/findings.go
package findings

type Severity string

const (
	Advisory Severity = "advisory"
	Blocking Severity = "blocking"
)

type Finding struct {
	File     string   `json:"file"`
	Line     int      `json:"line"`
	Rule     string   `json:"rule"`
	Severity Severity `json:"severity"`
	Message  string   `json:"message"`
	Source   string   `json:"source"`
}

type Result struct {
	Findings []Finding `json:"findings"`
}

func (r Result) ExitCode() int {
	code := 0
	for _, f := range r.Findings {
		if f.Severity == Blocking {
			return 2
		}
		if f.Severity == Advisory {
			code = 1
		}
	}
	return code
}

func (r Result) Count(sev Severity) int {
	n := 0
	for _, f := range r.Findings {
		if f.Severity == sev {
			n++
		}
	}
	return n
}

File: src/findings.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Advisory,
    Blocking,
}

#[derive(Debug, Clone, Serialize)]
pub struct Finding {
    pub file: String,
    pub line: usize,
    pub rule: String,
    pub severity: Severity,
    pub message: String,
    pub source: &'static str,
}

#[derive(Debug, Default, Serialize)]
pub struct Outcome {
    pub findings: Vec<Finding>,
}

impl Outcome {
    pub fn exit_code(&self) -> i32 {
        if self
            .findings
            .iter()
            .any(|f| f.severity == Severity::Blocking)
        {
            2
        } else if self.findings.is_empty() {
            0
        } else {
            1
        }
    }
}
